python/private/create_pkg_aliases.bzl

```diff
@@ -0,0 +1,139 @@
+
+    default_actual = "@{repo_name}_{version}_{dep}//:{target}".format(
+        repo_name = repo_name,
+        version = version_label(versions[-1]),
+        dep = dep,
+        target = target,
+    )
+    selects["//conditions:default"] = default_actual
+
+    return _SELECT.format(
+        name = name,
+        selects = "\n{}    ".format(
+            "".join([
+                "        {}: {},\n".format(repr(k), repr(v))
+                for k, v in selects.items()
+            ]),
+        ),
+    )
+
+def _render_entrypoints(repo_name, dep):
+    return """\
+load("@{repo_name}_{dep}//:entrypoints.bzl", "entrypoints")
+
+[
+    alias(
+        name = script,
+        actual = "@{repo_name}_{dep}//:" + target,
+        visibility = ["//visibility:public"],
+    )
+    for script, target in entrypoints.items()
+]
+""".format(
+        repo_name = repo_name,
+        dep = dep,
+    )
+
+def create_pkg_aliases(rctx, repo_name, bzl_packages, whl_map = None, rules_python = None):
+    """Create alias declarations for each python dependency.
+
+    The aliases should be appended to the pip_repository BUILD.bazel file. These aliases
+    allow users to use requirement() without needed a corresponding `use_repo()` for each dep
+    when using bzlmod.
+
+    Args:
+        rctx: the repository context.
+        repo_name: the repository name of the parent that is visible to the users.
+        bzl_packages: the list of packages to setup.
+        whl_map: the whl_map for generating Python version aware aliases.
+        rules_python: the name of the rules_python workspace.
+    """
+    for name in bzl_packages:
+        versions = None
+        if whl_map != None:
+            versions = whl_map[name]
+
+        build_content = "\n\n".join([
+            """package(default_visibility = ["//visibility:public"])""",
+            _render(versions = versions, name = name, repo_name = repo_name, dep = name, target = "pkg", rules_python = rules_python),
+        ] + [
+            _render(versions = versions, name = target, repo_name = repo_name, dep = name, target = target, rules_python = rules_python)
+            for target in ["pkg", "whl", "data", "dist_info"]
+        ])
+
+        rctx.file("{}/BUILD.bazel".format(name), build_content)
+
+        if versions == None:
+            build_content = _render_entrypoints(repo_name = repo_name, dep = name)
+            rctx.file("{}/bin/BUILD.bazel".format(name), build_content)
+            continue
+
+        # NOTE @aignas 2023-07-07: we are not creating aliases using a select
+        # and the version specific aliases because we would need to fetch the
+        # package for all versions in order to construct the said select.
+        for version in versions:
+            build_content = _render_entrypoints(repo_name = "{}_{}".format(repo_name, version_label(version)), dep = name)
+            rctx.file("{}/bin_py{}/BUILD.bazel".format(name, version_label(version)), build_content)
```

Comment:
I am wondering if this function should use `rctx.file` or if it should return a dict of files and their contents. It would allow for unit testing.

Reply:
I’m always for testing ;)